loader: attach XDP programs using bpf_link

rgo3 opened the pull request:
See individual commits.

The first two commits (prog removal and xdp bpf_link cleanup) will be submitted as a separate PR for backporting to 1.14 and 1.13 so these versions can handle downgrades from bpf_link-enabled Cilium agents. They were kept in this branch since future commits depended on them, and getting this broken up into smaller commits that make sense was already a challenge.

On a high level, this PR:
- attaches XDP progs using bpf_link
- adds per-device bpffs dirs, currently only for 'physical' devs, and only used for XDP links, at e.g. `/sys/fs/bpf/cilium/devices/enp5s0/links/cil_xdp_entry`
- adds a few utility functions to package bpf for working with bpffs and bpf_link pins
- refactors `attachProgram()` into `attachTCProgram()` and `attachXDPProgram()` (this was a minimal change)
- now uses ebpf-go's `link.XDPAttachFlags` as parameters instead of a simple u32
- adds netns-driven tests for attaching XDP progs to veths

This also puts us in a better position for adding tcx support during the next cycle.

Fixes: #27630

rgo3 committed:
loader: fix bug in obsolete XDP program removal

This commit fixes a bug where Cilium expected the XDP mode flag to be carried
in the netlink.LinkXdp.Flags field, while in reality the XDP mode is actually
contained in the LinkXdp.AttachMode field. Also added a regression test.

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>

asauber commented on pkg/datapath/loader/netlink_test.go:
```diff
@@ -446,32 +446,36 @@ func TestAttachProgram(t *testing.T) {
-	})
+	// the linkpath makes it a bit cumbersome to test this codepath of attachProgram now
+	// I think we can just remove it in favor of the new tests covering xdp attachements
+	// TODO(rgo3): remove or fix test
```
Given the tests added in this PR, my vote would be to remove this.

asauber commented on pkg/datapath/loader/xdp.go:
```diff
@@ -5,18 +5,30 @@ package loader
 
+// TODO: Clean up bpffs root logic and make this a var.
+func xdpLinkPath() string {
+	return filepath.Join(bpf.CiliumPath(), "loader", "links/xdp") // TODO(rgo3): not sure if we want that exact path, also relevant for cilium cleanup command
```
For now, these strings should at least be `const`s

asauber commented on cilium/cmd/cleanup.go:
```diff
@@ -582,6 +583,10 @@ func removeXDPs(links []netlink.Link) error {
 		if err != nil {
 			return err
 		}
+		err = os.Remove(filepath.Join(bpf.CiliumPath(), "loader", "links/xdp", link.Attrs().Name))
```
Declared `const`s can be used here rather than string literals.

asauber reviewed (changes requested):
Reviewed for Config and CLI. Didn't review the attach/detach logic.

rgo3 committed:
loader: clean up XDP bpf_links created by newer Cilium versions

A follow-up commit will introduce attaching XDP programs using bpf_link. Those
attachments cannot be overridden using netlink. If an older version of Cilium
wants to replace an XDP program on a managed interface, it'll need to remove
the bpf_link first.

This commit teaches the agent to remove the (currently) only XDP entrypoint,
cil_xdp_entry, before reattaching it using netlink. Note that this transition
is never seamless, since some time passes between deleting the link and
attaching the new program.

Also note that this downgrade path should seldom be used. If Cilium is upgraded
from a version with only netlink support, the new Cilium version will continue
to use netlink. Only if a fresh node is deployed with a new Cilium version
supporting bpf_link, then downgraded to one supporting only netlink, will the
bpf_link need to be removed.

For adding XDP bpf_link support, a per-device bpffs directory will be created
to pin the device's links, maps, etc. Follow-up commits will put XDP bpf_links
here. Subdirectories will be created for each device, for example:

/sys/fs/bpf/cilium/devices/enp5s0/

Then, following the convention we're already using for cgroup links, another
subdirectory will be created for links, followed by the entrypoint name:

/sys/fs/bpf/cilium/devices/enp5s0/links/cil_xdp_entry

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-authored-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: rename removeXDPs to removeXDPAttachments, unexport RemoveTCFilters

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: add MkdirBPF() function for creating bpffs dir hierarchies

We're using 0755 for all bpffs directories, so standardize on this.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: remove maybeRemoveXDPLinks

This was a forwards-compatibility measure intended for older agent versions
and is now redundant.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: return link.XDPAttachFlags from xdpModeToFlag

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
bpf: lift UpdateLink and UnpinLink helpers out of package socketlb

Package socketlb had existing code for updating and unpinning links
based on a pin path and a program handle. Factor these out into package
bpf so the XDP loader can start using them.

Signed-off-by: Timo Beckers <timo@isovalent.com>

rgo3 committed:
loader: attach and detach XDP programs using the bpf_link API

This extends the loader package to utilize bpf_link to attach XDP programs
if the kernel supports it. To avoid connectivity issues during upgrades,
existing XDP netlink attachments will not be replaced by bpf_link-based ones.
For these interfaces, we keep replacing the existing netlink XDP attachment.
bpf_link will only be used on newly-created or rebooted nodes. This matches
the behaviour of package socketlb.

A high-level overview of the changes:

- Introduced a simple loader.DetachXDP function to ensure a given link doesn't
  have a given XDP program attached. This takes care of detaching both netlink
  and xdp_link-based progs.
- attachProgram() was split into attachTCProgram() and attachXDPProgram().
- attachXDPProgram() implements the functionality outlined in the first
  paragraph.
- Tests were added for all code paths (clean host, existing netlink prog,
  as well as existing bpf_link prog).

Signed-off-by: Robin Gögge <r.goegge@isovalent.com>
Co-authored-by: Timo Beckers <timo@isovalent.com>

bimmlerd reviewed (approved):
CODEOWNERS change LGTM

dylandreimerink reviewed (approved):
LGTM!
